src/mame/f256/f256.cpp

```diff
@@ -0,0 +1,2011 @@
+        {
+            uint8_t src_byte = m_ram->read(src_addr + x + y * src_stride);
+            m_ram->write(dest_addr + x + y * dest_stride, src_byte);
+        }
+    }
+}
+void f256_state::performLinearDMA()
+{
+    uint32_t src_addr = ((m_iopage[0]->read(0xDF06) & 0x7) << 16) + (m_iopage[0]->read(0xDF05) << 8) + m_iopage[0]->read(0xDF04);
+    uint32_t dest_addr = ((m_iopage[0]->read(0xDF0A) & 0x7) << 16) + (m_iopage[0]->read(0xDF09) << 8) + m_iopage[0]->read(0xDF08);
+    uint32_t count = ((m_iopage[0]->read(0xDF0E) & 0x7) << 16) + (m_iopage[0]->read(0xDF0D) << 8) + m_iopage[0]->read(0xDF0C);
+    //logerror("Linear Copy DMA SRC: %X, DEST: %X, LEN: %X\n", src_addr, dest_addr, count);
+    memcpy(m_ram->pointer() + dest_addr, m_ram->pointer() + src_addr, count);
+}
+
+
+//-------------------------------------------------
+//  device_start
+//-------------------------------------------------
+void f256_state::device_start()
+{
+	driver_device::device_start();
+    reset_mmu();
+    // TODO: Copy the font from file to IO Page 1
+    //memcpy(m_iopage[1], m_font, 0x800);
+    for (int i=0;i<0x800;i++)
+    {
+        m_iopage[1]->write(i, m_font->as_u8(i));
+    }
+    // Copy the gamma correction table
+    uint8_t gamma_1_8[] = {
+        0x00, 0x0b, 0x11, 0x15, 0x19, 0x1c, 0x1f, 0x22, 0x25, 0x27, 0x2a, 0x2c, 0x2e, 0x30, 0x32, 0x34,
+        0x36, 0x38, 0x3a, 0x3c, 0x3d, 0x3f, 0x41, 0x43, 0x44, 0x46, 0x47, 0x49, 0x4a, 0x4c, 0x4d, 0x4f,
+        0x50, 0x51, 0x53, 0x54, 0x55, 0x57, 0x58, 0x59, 0x5b, 0x5c, 0x5d, 0x5e, 0x60, 0x61, 0x62, 0x63,
+        0x64, 0x65, 0x67, 0x68, 0x69, 0x6a, 0x6b, 0x6c, 0x6d, 0x6e, 0x70, 0x71, 0x72, 0x73, 0x74, 0x75,
+        0x76, 0x77, 0x78, 0x79, 0x7a, 0x7b, 0x7c, 0x7d, 0x7e, 0x7f, 0x80, 0x81, 0x82, 0x83, 0x84, 0x84,
+        0x85, 0x86, 0x87, 0x88, 0x89, 0x8a, 0x8b, 0x8c, 0x8d, 0x8e, 0x8e, 0x8f, 0x90, 0x91, 0x92, 0x93,
+        0x94, 0x95, 0x95, 0x96, 0x97, 0x98, 0x99, 0x9a, 0x9a, 0x9b, 0x9c, 0x9d, 0x9e, 0x9f, 0x9f, 0xa0,
+        0xa1, 0xa2, 0xa3, 0xa3, 0xa4, 0xa5, 0xa6, 0xa6, 0xa7, 0xa8, 0xa9, 0xaa, 0xaa, 0xab, 0xac, 0xad,
+        0xad, 0xae, 0xaf, 0xb0, 0xb0, 0xb1, 0xb2, 0xb3, 0xb3, 0xb4, 0xb5, 0xb6, 0xb6, 0xb7, 0xb8, 0xb8,
+        0xb9, 0xba, 0xbb, 0xbb, 0xbc, 0xbd, 0xbd, 0xbe, 0xbf, 0xbf, 0xc0, 0xc1, 0xc2, 0xc2, 0xc3, 0xc4,
+        0xc4, 0xc5, 0xc6, 0xc6, 0xc7, 0xc8, 0xc8, 0xc9, 0xca, 0xca, 0xcb, 0xcc, 0xcc, 0xcd, 0xce, 0xce,
+        0xcf, 0xd0, 0xd0, 0xd1, 0xd2, 0xd2, 0xd3, 0xd4, 0xd4, 0xd5, 0xd6, 0xd6, 0xd7, 0xd7, 0xd8, 0xd9,
+        0xd9, 0xda, 0xdb, 0xdb, 0xdc, 0xdc, 0xdd, 0xde, 0xde, 0xdf, 0xe0, 0xe0, 0xe1, 0xe1, 0xe2, 0xe3,
+        0xe3, 0xe4, 0xe4, 0xe5, 0xe6, 0xe6, 0xe7, 0xe7, 0xe8, 0xe9, 0xe9, 0xea, 0xea, 0xeb, 0xec, 0xec,
+        0xed, 0xed, 0xee, 0xef, 0xef, 0xf0, 0xf0, 0xf1, 0xf1, 0xf2, 0xf3, 0xf3, 0xf4, 0xf4, 0xf5, 0xf5,
+        0xf6, 0xf7, 0xf7, 0xf8, 0xf8, 0xf9, 0xf9, 0xfa, 0xfb, 0xfb, 0xfc, 0xfc, 0xfd, 0xfd, 0xfe, 0xff
+    };
+
+    memcpy(m_iopage[0]->pointer(), gamma_1_8, 256);
```

Comment:
What is this gamma table in hardware?

Reply:
The gamma table is used to correct the display colors.  When gamma is enabled in the display register, all the colors displayed are adjusted according to this indexed table.

Reply:
I understand what a gamma table is, I'm asking what hardware is involved in implementing it? Again, it's very unlikely that hardware is doing an arbitrary `memcpy()` from somewhere into RAM. Is it a RAMDAC or some other device with a fixed or programmable table that should be emulated?

Reply:
This is implemented with closed-source FPGA logic.